### ISSUE TITLE
Deadlock in v4.4.2

### ISSUE THREAD

**Issue by janekolszak** [laverdet/isolated-vm#330]
Hello!
We're experiencing a deadlock in **v4.4.2 + node 18.12.1**, but everything is OK with **v4.3.6 + node 16.18.0**

- strace for the locked process: https://drive.google.com/file/d/1PQAPCi3_zpHZ6ZlMIpTtN1hJcSUG8iP2/view?usp=share_link
- strace for everything when the lock occures:  https://drive.google.com/file/d/1oq4KCyYy4zAb9bKlI4EaGLXl9p3HO5HY/view?usp=share_link
- strace when everything is OK: https://drive.google.com/file/d/1jw3Gl4YCG-YZhFoPNE15p68YgBBn1AqT/view?usp=share_link

Thank you!

**Comment by laverdet** [comment]
Does v4.4.1 fail? Do you have a reproduction case I could run locally?

**Comment by ppedziwiatr** [comment]
Same issue in 4.4.1.

Anyway, we've decided to use a different tool, so this issue might be closed as well.

**Comment by kiddkai** [comment]
Hey @laverdet ,

Sorry for the delay, finally got a chance to get back to this issue, I was not able to reproduce it reliably, but here's a script that I can reproduce in node `16.8.1`. It will sometimes stuck at a specific `randombytes called x`. But again, not a reliable reproduction of the issue. I can only reproduce it randomly in MacOS(M1Pro). 

In macos+linux docker alpine image might be easier to reproduce this error.

The behaviour will be, run this script, wait for the counter stops moving. The log can pause in any time:

![image](https://user-images.githubusercontent.com/1309053/222633692-2eee314b-c2d7-4b31-a719-55d1b4d0ec63.png)


```js
const ivm = require('isolated-vm');
const crypto = require('crypto');

const isolate = new ivm.Isolate();

const task = async (j) => {
    const context = await isolate.createContext();
    await context.global.set("getProm", () => {
        const res = crypto.randomBytes(32);
        const { buffer } = res;

        console.log('randombyte called', j)

        return new ivm.ExternalCopy(buffer, {
            transferList: [buffer],
            transferOut: true
        }).copyInto({ release: true });
    }, { reference: true });
    await context.eval("getProm.applySync()")
    context.release();
}

const schedule = (n) => {
    const ts = [];
    for (let i = 0; i < 100; i++) {
        ts.push(task(n + i));
    }
    return ts
}

const run = async () => {
    for (let i = 0; i < Infinity; i++) {
        await Promise.all(schedule(i));
    }
}

run();
```

**Comment by kiddkai** [comment]
After the process stuck there, I pause the process and check the thread was waiting, it seem like for some reason it missed the conditional variable's notification from the other thread, so it stay there:

<img width="1612" alt="image" src="https://user-images.githubusercontent.com/1309053/222723340-427bf157-e174-4c45-ad8a-e2e591af1cc4.png">

**Comment by kiddkai** [comment]
Hi @laverdet,

I added an extra mutex and unique_lock it, and it seems to work. For some reason, I think the problem is caused by the wait not sharing the lock with `wake` and `ready`. I am not 100% sure by reading thru the cppreference thingy and examples about conditional_variables it seems like most of them are sharing the same mutex.

<img width="754" alt="image" src="https://user-images.githubusercontent.com/1309053/222860480-7d14a53c-8e9f-4d30-bc5a-c6ca1edbe2fa.png">

**Comment by kiddkai** [comment]
Arh, actually, in 4.4.2, the original deadlock is fixed, however. There's a new test case is failing. It seem like when main thread throws an error during execution, it will just hang there:

```js
const ivm = require('isolated-vm');
const crypto = require('crypto');

const isolate = new ivm.Isolate({
    memoryLimit: 2048
});
let c = 0;
const task = async (j) => {
    const context = await isolate.createContext();
    await context.global.set("getProm", () => {
        crypto.randomBytes(32);
        console.log('loop: ', c++);
        throw new Error('test');
    }, { reference: true });
    await context.eval("getProm.applySync()")
    context.release();
}

const schedule = (n) => {
    const ts = [];
    for (let i = 0; i < 100; i++) {
        ts.push(task(n + i));
    }
    return ts
}

const run = async () => {
    for (let i = 0; i < 1e5; i++) {
        await Promise.all(schedule(i));
    }
    isolate.dispose();
    console.log('pass')
}

run().catch(e => {
    console.log(e);
});
```

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/1309053/222991127-bf976635-3121-455f-b96f-d59a9bf1e581.png">

**Comment by kiddkai** [comment]
PR created for the fix: https://github.com/laverdet/isolated-vm/pull/346/files

It turns out it's a pretty simple one. The other synchronization bug introduced in 4.3.7 was fixed in 4.4.2, so this is just an additional patch to 4.4.2.

**Comment by kiddkai** [comment]
@laverdet I think should be safe to close this issue
